### PR TITLE
feat(transaction-annotation): Handle transactional context

### PR DESCRIPTION
**Pull request by prollandoc**

### CODE REVIEW

File: src/Proxy/ProxyGenerator/ServiceProxyGenerator.php [hunk 1]
```diff
@@ -57,17 +65,25 @@ public function generate(ReflectionClass $originalClass, ClassGenerator $classGe
...
-                    $additionalInterfaces['cache'] = ServiceProxyCacheInterface::class;
-                    $response = $this->cacheStrategy->execute(
-                        $this->serviceProxyStrategyRequestBuilder
-                            ->create()
-                            ->withAnnotation($methodAnnotation)
-                            ->withClass($originalClass)
-                            ->withMethod($method)
-                            ->build()
-                    );
+                if ($methodAnnotation instanceof ServiceProxyAnnotation) {
```
Comment (yassavic) [inline src/Proxy/ProxyGenerator/ServiceProxyGenerator.php:68]: I suggest to move this in a private method

File: README.md [hunk 2]
```diff
@@ -145,6 +148,52 @@ Supports Symfony ExpressionLanguage, for example:
...
+
+```php
+namespace MyProject\AClass;
+
+use OpenClassrooms\ServiceProxy\Annotations\Transaction;
+
+class AClass
+{
+    /**
+     * @Transaction(onConflictThrow="\MyProject\AClass\Exception\MyFunctionalException")
```
[the remark this answers is not in the document]
Comment (sidux) [reply README.md:192]: In your case, `onConflict` is based on code and an exception (PDO) so if you just have a mapping between and exception (PDO, Doctrine or whatever) and it's code to an other exception, you'll be agnostic and don't have to touch this library for future exceptions to handle, which is more flexible IMO.
for example : 
`exceptions=["\PDOException, 23000" => "\MyProject\AClass\Exception\MyFunctionalConflictException"]`
`exceptions=["\DoctrineConflictException" => "\MyProject\AClass\Exception\MyFunctionalConflictException"]`

You can even improve on this, and give to the transaction only a class containing the mapping which is implemented on our project to be more DRY.

**Comment by sidux** [comment]
Nice work @prollandoc ! 

There is something I'm asking my self and it's related to the introduction of attributes in PHP 8 which might change a lot of things in this library, I'd argue that our first priority (as we already discussed it in the past) is to upgrade our symfony and php version first, so that we won't end up changing a lot of things later.

Other then this we had old discussions on using libs like https://github.com/goaop/framework which already handles some of the complexity plus having a phpstorm plugin, and other bugs in this lib that we have to fix (https://github.com/OpenClassrooms/ServiceProxy/pull/6) so we are still not sure which direction this lib heads on.

Now that this is done if you think it doesn't need more effort we can try to merge it, but we have to be really careful with our testing.

**Comment by prollandoc** [comment]
@sidux Thanks! Yeah actually I was thinking that boosting the development of this library would allow us to autowire our use cases more easily. And you're right, it will have to be tested carefully. 
I'd like to handle those ordering issues before merging, and then we can go I guess. 

Regarding your previous comment about the mapping, it's a good idea and is easier to maintain, but I fear the configuration might be harder for the end user

**Comment by openclassrooms-admin** [comment]
Regarding priorities management, I do think it's an important issue:
we should not let someone break everything because of the annotation order.
(imagine a linter that change this order or someone that just forget)

If the transaction is not well placed, it could be a disaster.
And pre vs post vs exception could have different orders.
*Pre*
Security
Log
Event
Cache
Transaction

*Post*
Transaction
Cache
Event
Log

*Exception*
Transaction
Log
Event

If you put for example cache before the transaction in post, and the
transaction fails, it breaks the whole app.

Romain Kuzniak

Chief Technology Officer
[image: Openclassrooms Logo]

OpenClassrooms - 10 Quai de la Charente - 75019 Paris - France

Suivez-nous sur Twitter <https://twitter.com/openclassrooms>, Facebook
<https://www.facebook.com/openclassrooms>, LinkedIn
<https://www.linkedin.com/school/openclassrooms/mycompany/>

**Comment by romainkuzniak** [comment]
And for the AOP, aop is a performance killer in PHP. We used to do it, and never again. Because it implies file reading at runtime. And real proxy doesn't exist in PHP. That's why we need to write file (and simulate Proxy)
Now, I haven't checked if PHP8 provides better mechanisms.

**Comment by sidux** [comment]
@romainkuzniak That lib creates proxies also it just does it on the composer autoload mechanism so it's not dependent on a class being a service (but it's main issue is that it's not actively maintained anymore, just bug fixing)

**Pull request closed without merging**